src/components/tl/cuda/bcast/bcast_linear.c

```diff
@@ -0,0 +1,303 @@
+                  (task->bcast_linear.step - 1) * half_scratch_size;
+    size_t offset_buff = task->bcast_linear.step * half_scratch_size;
+
+    ucc_ee_executor_t *     exec;
+    ucc_ee_executor_task_t *etask;
+    ucc_status_t            st;
+    void *                  sbuf, *dbuf;
+    int                     i;
+
+    task->super.status = UCC_INPROGRESS;
+
+    st = ucc_coll_task_get_executor(&task->super, &exec);
+    if (ucc_unlikely(st != UCC_OK)) {
+        return;
+    }
+
+    switch (task->bcast_linear.stage) {
+    case STAGE_SYNC:
+        if (ucc_tl_cuda_get_sync(task) != UCC_OK) {
+            task->super.status = UCC_INPROGRESS;
+            return;
+        }
+        task->bcast_linear.step = 0;
+        st                      = ucc_tl_cuda_bcast_linear_setup_start(task);
+        if (st != UCC_OK) {
+            task->super.status = st;
+            return;
+        }
+        task->bcast_linear.stage = STAGE_SETUP;
+    case STAGE_SETUP:
+        st = ucc_tl_cuda_bcast_linear_setup_test(task);
+        if (st != UCC_OK) {
+            task->super.status = st;
+            return;
+        }
+        ucc_tl_cuda_put_sync(task);
+        if (trank == task->bcast_linear.root) {
+            task->bcast_linear.stage = STAGE_COPY;
+        } else {
+            task->bcast_linear.stage = STAGE_WAIT_ROOT;
+        }
+    default:
+        break;
+    }
+
+    if (trank == task->bcast_linear.root) {
+        // Root scenario
+        // fall-through between cases is intentional
+        switch (task->bcast_linear.stage) {
+        case STAGE_COPY:
+            // copy from src buffer to scratch
+            dbuf = PTR_OFFSET(TASK_SCRATCH(task, trank),
+                              task->bcast_linear.step % 2 * half_scratch_size);
+            sbuf = PTR_OFFSET(task->bcast_linear.sbuf, offset_buff);
+            st   = ecopy(dbuf, sbuf, chunk_size, exec,
+                       &task->bcast_linear.exec_task);
+            if (st != UCC_OK) {
+                ucc_error("failed to post ecopy task");
+                task->super.status = st;
+                return;
+            }
+            task->bcast_linear.stage = STAGE_WAIT_COPY;
+        case STAGE_WAIT_COPY:
+            etask = task->bcast_linear.exec_task;
+            if (etask) {
+                st = ucc_ee_executor_task_test(etask);
+                if (st == UCC_OK) {
+                    ucc_ee_executor_task_finalize(etask);
+                    task->bcast_linear.exec_task = NULL;
+                    // signal others
+                    ++task->bcast_linear.step;
+                    set_rank_step(task, task->bcast_linear.root,
+                                  task->bcast_linear.step, 0);
+                    task->bcast_linear.stage = STAGE_WAIT_ALL;
+                } else {
+                    // not ready
+                    return;
+                }
+            } else {
+                ucc_debug("etask is nullptr");
```

Comment:
isn't this case an infinite loop? I am not sure to understand

Reply:
it is error case, used it for debug, it should not happen in real situations

Reply:
Ok, why not using `ucc_assert` here then?